File: abc/181/e.cpp
#include <bits/stdc++.h>
#define _USE_MATH_DEFINES  // M_PI等のフラグ
#define MOD 1000000007
#define COUNTOF(array) (sizeof(array)/sizeof(array[0]))
#define rep(i,n) for (int i = 0; i < (n); ++i)
using namespace std;
using ll = long long;
using P = pair<int,int>;


void solve(){
    int N, M;
    cin >> N >> M;
    vector<int> H(N);
    vector<int> W(M);
    for (int i=0; i<N; i++) cin >> H[i];
    for (int i=0; i<M; i++) cin >> W[i];
    sort(H.begin(), H.end());
    sort(W.begin(), W.end());

    /*
    i=0から順に、先生と組む場合の身長差を計算をする
    i=1のときは、しゃくとり法を使えば先生と組む以外の人たちの身長差は一瞬で計算できるO(1)
    先生がどの身長になればよいかは、二分探索で求めればO(logN)
    よって、この計算量はO(NlogN)でいけそう
    */
    vector<int> diffT(N,0);  // 生徒iと、先生との最小の身長差
    for (int i=0; i<N; i++) {
        auto it = lower_bound(W.begin(), W.end(), H[i]);
        if (it == W.begin()) {
            diffT[i] = W[0]-H[i];
        }
        else if (it == W.end()) {
            diffT[i] = H[i]-W.back();
        }
        else {
            diffT[i] = min(abs(H[i]-*it), abs(H[i]-*(it-1)));
        }
    }

    // i=0の生徒が先生とペアになるとき
    int total = 0;
    for (int i=1; 2*i<N; i++) {
        total += H[2*i]-H[2*i-1];
    }
    int ans = total + diffT[0];

    // i=1~N-1の生徒が先生とペアになるときを、しゃくとり法で求める
    for (int i=1; i<N; i++) {
        if (i%2==1) {
            total -= H[i+1] - H[i];
            total += H[i+1] - H[i-1];
            ans = min(ans, total+diffT[i]);
        }
        else {
            total -= H[i] - H[i-2];
            total += H[i-1] - H[i-2];
            ans = min(ans, total+diffT[i]);
        }
    }

    // 出力
    cout << ans << endl;
}


int main(int argc, char const *argv[]){
    solve();
    return 0;
}

File: abc/181/e.py
# -*- coding:utf-8 -*-

def solve():
    import bisect

    N, M = list(map(int, input().split()))
    H = list(map(int, input().split()))
    W = list(map(int, input().split()))
    H.sort()
    W.sort()

    # 生徒iが先生と組むときの最小身長差を求める
    diffs = [0 for _ in range(N)]
    for i in range(N):
        j = bisect.bisect_left(W, H[i])
        if j == 0:
            diffs[i] = abs(H[i]-W[0])
        elif j == M:
            diffs[i] = abs(H[i]-W[-1])
        else:
            diffs[i] = min(abs(H[i]-W[j-1]), abs(H[i]-W[j]))

    # まず、H[0]が先生と組むときを計算する
    total = 0
    for i in range(1, len(H), 2):
        total += H[i+1]-H[i]
    ans = total+diffs[0]

    # 次に、H[1]~が先生と組むときの計算をする（しゃくとり法で）
    for i in range(1, N):
        if i%2==1:
            total += H[i]-H[i-1]
        else:
            total -= H[i]-H[i-1]
        ans = min(ans, total+diffs[i])

    print(ans)


if __name__ == "__main__":
    solve()
